stdlib/public/core/UnsafeRawBufferPointer.swift.gyb

```diff
@@ -0,0 +1,386 @@
+    self.init(start: bytes.baseAddress, count: bytes.count)
+  }
+%  end # !mutable
+
+  /// Creates an `${Self}` over the contiguous bytes in `buffer`.
+  ///
+  /// - Precondition: `T` is a trivial type.
+  public init<T>(_ buffer: UnsafeMutableBufferPointer<T>) {
+    self.init(start: buffer.baseAddress!,
+      count: buffer.count * MemoryLayout<T>.stride)
+  }
+
+%  if not mutable:
+  /// Creates an `${Self}` view over the contiguous memory in `buffer`.
+  ///
+  /// - Precondition: `T` is a trivial type.
+  public init<T>(_ buffer: UnsafeBufferPointer<T>) {
+    self.init(start: buffer.baseAddress!,
+      count: buffer.count * MemoryLayout<T>.stride)
+  }
+%  end # !mutable
+
+  /// Always zero, which is the index of the first byte in a
+  /// non-empty buffer.
+  public var startIndex: Int {
+    return 0
+  }
+
+  /// The "past the end" position---that is, the position one greater than the
+  /// last valid subscript argument.
+  ///
+  /// The `endIndex` property of an `Unsafe${Mutable}RawBufferPointer`
+  /// instance is always identical to `count`.
+  public var endIndex: Int {
+    return count
+  }
+
+  public typealias Indices = CountableRange<Int>
+
+  public var indices: Indices {
+    return startIndex..<endIndex
+  }
+
+  /// Accesses the `i`th byte in the memory region as a `UInt8` value.
+  public subscript(i: Int) -> UInt8 {
+    get {
+      _debugPrecondition(i >= 0)
+      _debugPrecondition(i < endIndex)
+      return _position!.load(fromByteOffset: i, as: UInt8.self)
+    }
+%  if mutable:
+    nonmutating set {
+      _debugPrecondition(i >= 0)
+      _debugPrecondition(i < endIndex)
+      _position!.storeBytes(of: newValue, toByteOffset: i, as: UInt8.self)
+    }
+%  end # mutable
+  }
+
+  /// Accesses the bytes in the memory region within `bounds` as a `UInt8`
+  /// values.
+  public subscript(bounds: Range<Int>) -> Unsafe${Mutable}RawBufferPointer {
+    get {
+      _debugPrecondition(bounds.lowerBound >= startIndex)
+      _debugPrecondition(bounds.upperBound <= endIndex)
+      return Unsafe${Mutable}RawBufferPointer(
+        start: baseAddress.map { $0 + bounds.lowerBound },
+        count: bounds.count)
+    }
+%  if mutable:
+    set {
+      _debugPrecondition(bounds.lowerBound >= startIndex)
+      _debugPrecondition(bounds.upperBound <= endIndex)
+      _debugPrecondition(bounds.count == newValue.count)
+
+      if newValue.count > 0 {
+        (baseAddress! + bounds.lowerBound).copyBytes(
+          from: newValue.baseAddress!,
+          count: newValue.count)
+      }
+    }
+%  end # mutable
+  }
+
+  /// Returns an iterator over the bytes of this sequence.
+  ///
+  /// - Complexity: O(1).
+  public func makeIterator() -> Iterator {
+    return Iterator(_position: _position, _end: _end)
+  }
+
+  /// A pointer to the first byte of the buffer.
+  public var baseAddress: Unsafe${Mutable}RawPointer? {
+    return _position
+  }
+
+  /// The number of bytes in the buffer.
+  public var count: Int {
```

Comment:
Well, you have to define one in terms of the other I guess.  The difference is that count has a default implementation so if we made them both store a pair of pointers instead of a pointer + length we'd get it with less code.